SiFive feedback (#29)
riscv-non-isa/riscv-security-model: specification/src/chapter3.adoc

=== Isolation

> Isolation enables the separation of software components executing on a Hart, as well
> as separate device assignment. RISC-V enables:

Reviewer: ```suggestion
Isolation enables access restrictions on software components executing on a hart, as well
as on device accesses. RISC-V enables:
```

NOTE: IOMTT can also be sufficient for protecting Root devices in the sense that
M-mode can enforce that its own resources are never assigned to another domain.
Use of IOPMP or similar still adds further protections. For example, a system

> could require that Root devices are not able to access memory assigned to TEE

Reviewer: ```suggestion
may require that Root devices are not able to access memory assigned to TEE
```

=== Software enforced memory tagging

*See the https://github.com/riscv/riscv-j-extension[RISC-V Pointer Masking] specification.*

_Memory tagging (MT)_, is a technique which can improve the memory safety of an
application. A part of the effective address of a pointer can be masked off
and used as a tag indicating the intended ownership or state of a pointer. The tag
can be used to track accesses across different regions as well as protecting
against pointer misuse such as "use-after-free". Pointer masking implementations should use
the proposed RISC-V pointer masking extension (Smmpm, Smnpm, Ssnpm).

With software based memory tagging the access rules encoded in tags are

> enforced by software, such as the compiler and the C library.

Reviewer: ```suggestion
enforced by software, such as the compiler and the application runtime.
```

CHERI - an ISA technique that uses capability-based memory protection for spatial and temporal memory safety, compartmentalization, and control-flow enforcement.  Source code has to be recompiled to capture memory safety properties inherent in the source language.

===== Capstone

*See the https://github.com/project-starch/capstone-spec[Capstone-RISCV ISA Reference].*

Capstone is another ISA technique aiming to enable a capability architecture in software and hardwre.

==== Hardware enforced memory tagging

Hardware enforced memory tagging extends software based memory tagging (see
above) by moving enforcement of tagged rules to hardware. At present this technology is at a very early
stage of discussion for RISC-V.

==== Hardware-assisted Fault Isolation

_Hardware-assisted fault isolation (HFI)_ aims to provide lightweight in-process isolation to mitigate against errors
in one process compromising other processes within the same workload.

==== Lightweight isolation

Lighweight isolation aims to enable the isolation of M-Mode runtimes and libraries. At present this technology is at a very early
stage of discussion for RISC-V.

==== System integration

> The _WorldGuard security model_ provides an open, system level approach to securing access to system wide resources by
> by software applications. The WorldGuard model introduces a concept of Worlds with unique World IDs. Software running at
> different RISC-V privilege levels can be assigned unique World IDs. System resources can be associated with lists of
> World IDs to control accesses to those resources in a fine grained manner.

Reviewer: ```suggestion
WorldGuard introduces a concept of World IDs, which may be uniquely and statically-assigned
to software running at RISC-V privilege levels . System resources can be then be associated with
World IDs in order to control accesses to those resources in a fine-grained manner.
```

==== Cryptography enhancements

The _RISC-V Post Quantum Cryptography_ initiative aims to specify ISA extensions that enhance performance and
implementation efficiency for contemporary public-key cryptography, with a focus on standard Post-Quantum Cryptography

> algorithms like Kyber, Dilithium, and others. The ISA design and evaluation prioritize the requirements of real-world

Reviewer 2: I can't recall if this was previously in the document, but we should definitely use the FIPS standards to refer to the standardized PQC algorithms. In this case it would be FIPS-203 (Kyber)/204 (Dilithium)/205 (SPHINCS+). Moreover, the standardized algorithm are not an exact 1:1 match to the "original" algorithms.

Author: Thanks @lfiolhais.

Would the following text be OK as a replacement for the relevant snippet ?

```
The _RISC-V Post Quantum Cryptography_ initiative aims to specify ISA extensions that enhance performance and
implementation efficiency for contemporary public-key cryptography, with a focus on standard Post-Quantum Cryptography
algorithms like FIPS-203 (Kyber), FIPS-24(Dilithium), FIPS-205(SPHINCS+) and others.
```

Reviewer 3: Thanks, makes sense.

Reviewer 2: Great! We also need the citations ;)

Reviewer 4: Hi @lfiolhais.

Please note: IANACG (I am not a cryptography expert!) :) Please bear with me.

I'm looking for URLs to include as citations in Appendix A. Here's what I have, please could you approve ?

1. FIPS-203 (Kyber): https://csrc.nist.gov/pubs/fips/203/ipd
- I couldn't confirm whether the string 'Kyber' is synonymous with FIPS-203

2. FIPS-24 (Dilithium): I couldn't find anything.
- Is it FIPS-24 or FIPS-204 ?
- If it is FIPS-204, then is https://csrc.nist.gov/pubs/fips/204/ipd ok to use as the URL ?
- Note that I couldn't confirm whether the string 'Dilithium' is synonymous with FIPS-204. Please can you confirm ?

3. FIPS-205 (SPHINCS+): https://csrc.nist.gov/pubs/fips/205/ipd

Thanks!

Reviewer 2: You have the correct links for all of them! Kyber ~ FIPS-203. Dilithium ~ FIPS-204. SPHINCS+ ~ FIPS-205. Technically speaking, neither of the FIPS standards are synonymous with their counterparts. They have subtle differences between the original algorithms and the FIPS standards. I'm also not a cryptographer but I've been told from multiple cryptographers that they aren't one and the same.

==== System integration

WorldGuard introduces a concept of World IDs, which are unique system-wide identifiers.

> The World Ids may be statically-assigned to RISC-V hardware, and thus software at various

Reviewer: ```suggestion
The World Ids may be statically-assigned to RISC-V hardware, and thus to software operating at various
```
